common/usermode.c

```diff
@@ -111,6 +114,26 @@ void init_usermode(percpu_t *percpu) {
 
     init_syscall();
     init_sysenter();
+    set_intr_gate(&percpu->idt[SYSCALL_INT], __KERN_CS, _ul(int80_handler_entry),
+                  GATE_DPL3, GATE_PRESENT, 0);
+}
+
+static inline long __user_text _int80(long syscall_nr, long arg1, long arg2, long arg3,
+                                      long arg4, long arg5) {
+    register long return_code asm(STR(_ASM_AX));
+    register long _arg4 asm("r8") = arg4;
+    register long _arg5 asm("r9") = arg5;
+
+    /* clang-format off */
+    asm volatile(
+        "int $" STR(SYSCALL_INT) "\n"
+        : "=a"(return_code)
+        : "0"(syscall_nr), "S"(arg1), "d"(arg2), "D" (arg3), "r"(_arg4), "r"(_arg5)
+        : STR(_ASM_CX), "r11"
```

Comment:
this clobbering is not needed

Reply:
You're right, but looking at the syscall handlers I can see that we don't guarantee that (even `syscall_handler_entry` is wrong in this regard!).

By making use of `SAVE_CALLEE_SAVED_REGS` we actually save the _wrong_ set of registers -- the ones already preserved by called C functions. The callee clobbered ones will be, well, clobbered.

It should either be `SAVE_ALL_REGS`, for simplicity, or we need to introduce a `SAVE_CALLEE_CLOBBERED_REGS` macro that takes care of preserving the callee clobbered registers.

Reply:
Oh yes, I agree. I would go with the `SAVE_ALL_REGS` for the simplicity in all the handlers.

Reply:
@sktt, as you're touching the code already, can you fix that as well? For the sysenter and int80 handlers added in this series directly, and for the syscall handler with a separate commit?